tools/cabana/chart/chart.cc

```diff
@@ -599,20 +602,19 @@ void ChartView::startAnimation() {
 
 void ChartView::paintEvent(QPaintEvent *event) {
   if (!can->liveStreaming()) {
+    const qreal dpr = viewport()->devicePixelRatioF();
     if (chart_pixmap.isNull()) {
-      const qreal dpr = viewport()->devicePixelRatioF();
       chart_pixmap = QPixmap(viewport()->size() * dpr);
-      chart_pixmap.setDevicePixelRatio(dpr);
       QPainter p(&chart_pixmap);
       p.setRenderHints(QPainter::Antialiasing);
-      drawBackground(&p, viewport()->rect());
+      drawBackground(&p, QRect(viewport()->rect().topLeft(), viewport()->rect().size() * dpr));
       scene()->setSceneRect(viewport()->rect());
       scene()->render(&p);
```

Comment:
It is also possible that there is no problem anywhere else, except this line.

Reply:
should we set target and source rect  ? https://doc.qt.io/qt-6/qgraphicsscene.html#render